### PR TITLE
Add namespacing to pkgdb

### PR DESCRIPTION
**Pull request by pypingou**
This will allow pkgdb to manage git repos for rpms but also other repos such as
docker or anything else.

### CODE REVIEW

File: pkgdb2/api/extras.py [hunk 1]
```diff
@@ -158,18 +160,25 @@ def _vcs_acls_cache(out_format='text', eol=False, collection=None):
...
         oformat=out_format,
-        skip_pp=APP.config.get('PKGS_NOT_PROVENPACKAGER', None))
+        skip_pp=APP.config.get('PKGS_NOT_PROVENPACKAGER', None),
+        namespace=namespace)
     output = []
     if out_format == 'json':
-        output = {'packageAcls': packages,
-                  'title': 'Fedora Package Database -- VCS ACLs'}
+        output = packages
+        if 'rpms' in output:
```
Comment (ralphbean) [inline pkgdb2/api/extras.py]: Can we get away without special-casing this in the pkgdb server?

What reads this JSON endpoint -- the script that generates ACLs for gitolite?  I presume this change is for backwards compatibility.  By special casing this here on the pkgdb side, we'll have to *also* add special-case logic to that ACL generation script.

Can we keep the namespacing in the JSON response (to the detriment of backwards compatibility), please, and keep this stretch of code ignorant of the namespace details?

File: pkgdb2/api/extras.py [hunk 2]
```diff
@@ -158,18 +160,25 @@ def _vcs_acls_cache(out_format='text', eol=False, collection=None):
...
+        output['title'] = 'Fedora Package Database -- VCS ACLs'
     else:
         for package in sorted(packages):
             for branch in sorted(packages[package]):
                 if packages[package][branch]['group']:
                     packages[package][branch]['group'] += ','
                 output.append(
-                    'avail | %(group)s%(user)s | rpms/%(name)s/%(branch)s'
+                    'avail | %(group)s%(user)s | '
+                    '%(namespace)s/%(name)s/%(branch)s'
```
Comment (ralphbean) [inline pkgdb2/api/extras.py]: Cool!

File: pkgdb2/api/packages.py [hunk 3]
```diff
@@ -1287,9 +1291,11 @@ def api_package_request():
             cnt += 1
         collections.append(collection)
 
+    namespaces = pkgdblib.get_status(SESSION, 'namespaces')['namespaces']
```
Comment (ralphbean) [inline pkgdb2/api/packages.py]: Seems like this should be singular, i.e. ``namespace`` not plural, i.e. ``namespaces``.  Here, and everywhere, no?

There's never a time that a package could have multiple namespaces, correct?

File: pkgdb2/lib/__init__.py [hunk 4]
```diff
@@ -2579,3 +2602,88 @@ def get_retired_packages(session, collection):
...
+        return 'Namespace "%s" created' % namespace
+    except SQLAlchemyError, err:  # pragma: no cover
+        pkgdb2.LOG.exception(err)
+        session.rollback()
+        raise PkgdbException(
+            'Could not add Namespace "%s" to the database.' % namespace)
+
+
+def drop_namespace(session, namespace, user):
+    """ Remove a namespace to the database.
```
Comment (ralphbean) [inline pkgdb2/lib/__init__.py]: ``s/to the database/from the database/``

File: pkgdb2/templates/list_packages.html [hunk 5]
```diff
@@ -33,7 +33,7 @@
 
 {% block content %}
 
-<h1>Search packages</h1>
+<h1>Search packages{% if namespace %} ({{ namespace }}) {% endif %}</h1>
```
Comment (ralphbean) [inline pkgdb2/templates/list_packages.html]: Nice!

File: pkgdb2/templates/master.html [hunk 6]
```diff
@@ -47,8 +47,13 @@
               <option value="retired"
               {% if select=='retired' %} selected {% endif %}>
                   Retired</option>
+              {% for ns in namespaces %}
+              <option value="{{ ns }}"
+              {% if select==ns %} selected {% endif %}>
```
Comment (ralphbean) [inline pkgdb2/templates/master.html]: Should this be ``if select=='rpms'`` ?  The default?

Comment (ralphbean) [inline pkgdb2/templates/master.html]: Or, no... I'm confused by this block, I guess.  Something seems off.

File: pkgdb2/templates/package.html [hunk 7]
```diff
@@ -200,8 +200,13 @@
 {% block content %}
 
 <section id="pkg_info">
-  <h1 class="inline" property="doap:name">{{ package.name }}</h1>{% if package.upstream_url
-      %} (<a class="inline" href="{{ package.upstream_url }}">upstream</a>){% endif %}
+  <h1 class="inline">
+      <span class="italic">{{ package.namespace | capitalize}}:</span>
```
Comment (ralphbean) [inline pkgdb2/templates/package.html]: Replace the ``:`` with ``/``?  We want to encourage people to think of the namespaces as directories everywhere I think (that was what I got out of the devel list discussion).  I.e.:

- rpms/nethack
- rpms/cockpit

and

- dockerfiles/cockpit

In this section of the template, can we ensure there is no space inbetween the ``/`` and the text on either side, too?

File: pkgdb2/api/packages.py (hunk 3 above)
Comment (ralphbean) [inline pkgdb2/api/packages.py]: I see now -- this is getting all of the namespaces in order to populate a drop-down menu, correct?  If that is correct, then this plural usage is also correct.  (my bad).

File: pkgdb2/api/extras.py (hunk 1 above)
Comment (pypingou) [inline pkgdb2/api/extras.py]: It was for backward compatibility but I'm fine with breaking it. It's an ``extra`` API endpoint anyway and we haven't promise to keep it stable.

File: pkgdb2/api/packages.py (hunk 3 above)
Comment (pypingou) [inline pkgdb2/api/packages.py]: It's the idea yes (getting all of them from the DB to populate the drop-down list)

File: pkgdb2/templates/master.html (hunk 6 above)
Comment (pypingou) [inline pkgdb2/templates/master.html]: select is provided to the template and here, in the options for searching pkgdb, we try to make sure that if you searched for ``packagers`` or ``docker``, the drop-down of the search options is set to what you searched (``packagers`` or ``docker``)

File: pkgdb2/templates/package.html (hunk 7 above)
Comment (pypingou) [inline pkgdb2/templates/package.html]: ok with the ``:`` -> ``/`` move, I'll check for the space.

File: pkgdb2/templates/master.html (hunk 6 above)
Comment (ralphbean) [inline pkgdb2/templates/master.html]: Cool, I understand now.

**Comment by pypingou** [comment]
Tests are broken currently, will fix them shortly

**Comment by ralphbean** [comment]
Cool.  This looks good.

Aside from fixing the VCS regeneration script to handle the new output.. can you think of any other systems that might need to be modified to handle this stuff?

**Comment by pypingou** [comment]
No that's the only one I can think of right now

**Comment by pypingou** [comment]
Rebased and adjusted :)

**Comment by ralphbean** [comment]
OK, :+1: here (assuming we can adjust the gitolite ACL script at release/deploy time).

**Comment by ralphbean** [comment]
Time to merge, no?  :+1: here.  :)

**Comment by pypingou** [comment]
First thing on the list for tomorrow, release 2.0 :)

**Pull request merged**